FormalConjectures/Paper/DegreeSequencesTriangleFree.lean

```diff
@@ -0,0 +1,173 @@
+
+/-- The number of vertices of `G` having degree `d`. -/
+noncomputable def degreeFreq (G : SimpleGraph α) (d : ℕ) : ℕ :=
+  (Finset.univ.filter fun v : α => G.degree v = d).card
+
+/-- `δ G` is the minimum degree of `G`. -/
+noncomputable def δ (G : SimpleGraph α) [DecidableRel G.Adj] : ℕ :=
+  sInf (Set.range fun v : α => G.degree v)
+
+end SimpleGraph
+
+section lemmas
+
+variable (d : ℕ → ℕ) (n k r : ℕ)
+
+/-- **Lemma 1 (a)** -/
+@[category research solved, AMS 5]
+lemma lemma1_a
+    (h_no_three : ∀ k, d (k + 2) ≠ d k) :
+    1 ≤ d (k + 2) - d k := by
+  sorry
+
+/-- **Lemma 1 (b)** -/
+@[category research solved, AMS 5]
+lemma lemma1_b
+    (h_no_three : ∀ i, d (i + 2) ≠ d i) :
+    r ≤ d (k + 2 * r) - d k := by
+  sorry
+
+/-- **Lemma 2 (a)** -/
+@[category research solved, AMS 5]
+lemma lemma2_a
+    (h_no_three : ∀ i, d (i + 2) ≠ d i) :
+    2 * n * n ≤
+      (∑ i ∈ Finset.Icc (2 * n + 1) (4 * n), d i) -
+        (∑ i ∈ Finset.Icc 1 (2 * n), d i) := by
+  sorry
+
+/-- **Lemma 2 (b)** -/
+@[category research solved, AMS 5]
+lemma lemma2_b
+    (h_no_three : ∀ i, d (i + 2) ≠ d i) :
+    2 * n * n + 2 * n + 1 ≤
+      (∑ i ∈ Finset.Icc (2 * n + 1) (4 * n + 1), d i) -
+        (∑ i ∈ Finset.Icc 1 (2 * n), d i) := by
+  sorry
+
+/-- **Lemma 2 (c)** -/
+@[category research solved, AMS 5]
+lemma lemma2_c
+    (h_no_three : ∀ i, d (i + 2) ≠ d i) :
+    2 * n * n + 2 * n ≤
+      (∑ i ∈ Finset.Icc (2 * n + 2) (4 * n + 2), d i) -
+        (∑ i ∈ Finset.Icc 1 (2 * n + 1), d i) := by
+  sorry
+
+/-- **Lemma 2 (d)** -/
+@[category research solved, AMS 5]
+lemma lemma2_d
+    (h_no_three : ∀ i, d (i + 2) ≠ d i) :
+    2 * n * n + 4 * n + 2 ≤
+      (∑ i ∈ Finset.Icc (2 * n + 2) (4 * n + 3), d i) -
+        (∑ i ∈ Finset.Icc 1 (2 * n + 1), d i) := by
+  sorry
+
+end lemmas
+
+namespace SimpleGraph
+
+variable {α : Type*} [Fintype α] [DecidableEq α]
+
+/-- The degree sequence of a graph, sorted in nondecreasing order. -/
+noncomputable def degreeSequence (G : SimpleGraph α) [DecidableRel G.Adj] : List ℕ :=
+  (Finset.univ.val.map fun v : α => G.degree v).sort (· ≤ ·)
+
+/-- The degree sequence of `G` is **compact** if it satisfies
+`IsCompactSequence`. -/
+def degreeSequenceCompact (G : SimpleGraph α) [DecidableRel G.Adj] : Prop :=
+  IsCompactSequence fun k => (degreeSequence G).getD k 0
+
+/-- **Theorem 1.** If a triangle-free graph has `f = 2`,
+then it is bipartite, has minimum degree `1`, and
+its degree sequence is compact. -/
+@[category research solved, AMS 5]
+theorem theorem1 (G : SimpleGraph α) [DecidableRel G.Adj] (h₁ : G.CliqueFree 3) (h₂ : f G = 2) :
+    G.IsBipartite ∧ δ G = 1 ∧ degreeSequenceCompact G := by
+  sorry
+
+/-- **Lemma 3.** For every `n` there exists a bipartite graph with
+`8 n` vertices, minimum degree `n + 1`, and `f = 3`. -/
+@[category research solved, AMS 5]
+lemma lemma3 (n : ℕ) :
+    ∃ (G : SimpleGraph (Fin (8 * n))) (_ : DecidableRel G.Adj),
+      G.IsBipartite ∧ δ G = n + 1 ∧ f G = 3 := by
```

Comment:
I think you'll want `n` to be positive here!

Reply:
Done, thank you!